### ISSUE TITLE
JSON error

### ISSUE THREAD

**Issue by mcgurdan** [domwillcode/yale-smart-alarm-client#5]
Hi, after updating to the latest code I now get the following error when running a very simple python script:

Script:
from yalesmartalarmclient.client import YaleSmartAlarmClient

username = "myEmailAddressHere"
password = "myPasswordHere"
client = YaleSmartAlarmClient(username,password)
client.get_armed_status()

Error:

pi@raspberrypi:~ $ 
pi@raspberrypi:~ $ python  alarm_part_arm.py
Traceback (most recent call last):
  File "alarm_part_arm.py", line 6, in <module>
    client.get_armed_status()
  File "/usr/local/lib/python2.7/dist-packages/yalesmartalarmclient/client.py", line 49, in get_armed_status
    alarm_state = self._get_authenticated(self._ENDPOINT_GET_MODE)
  File "/usr/local/lib/python2.7/dist-packages/yalesmartalarmclient/client.py", line 97, in _get_authenticated
    data = response.json()
  File "/usr/local/lib/python2.7/dist-packages/requests/models.py", line 892, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/simplejson/__init__.py", line 518, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python2.7/dist-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/local/lib/python2.7/dist-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

**Comment by laf** [comment]
That's really odd, I've taken your test script and it runs fine for me. Even tried on python 2.7

You could edit client.py and add this above line 97:

`print(response)`

Re-run it and see what the output is showing.

**Comment by mcgurdan** [comment]
Thanks I have done so, please see below, looks to me it just printing the HTTP status code:

<Response [200]>
Traceback (most recent call last):
  File "alarm_part_arm.py", line 6, in <module>
    client.get_armed_status()
  File "/usr/local/lib/python2.7/dist-packages/yalesmartalarmclient/client.py", line 49, in get_armed_status
    alarm_state = self._get_authenticated(self._ENDPOINT_GET_MODE)
  File "/usr/local/lib/python2.7/dist-packages/yalesmartalarmclient/client.py", line 98, in _get_authenticated
    data = response.json()
  File "/usr/local/lib/python2.7/dist-packages/requests/models.py", line 892, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/simplejson/__init__.py", line 518, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python2.7/dist-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/local/lib/python2.7/dist-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

**Comment by mcgurdan** [comment]
I also added the print to the next response and noticed I get a 403 http response. Forbidden.

Any ideas what would cause this?

**Comment by mcgurdan** [comment]
Could you send bad creds and see if you get the same 403, although my credentials are identical to what I use to login to the app and that is working fine. There isn’t a pre-req that I need to login to some web service or something beforehand?

**Comment by domwillcode** [comment]
This is odd. It's working fine standalone for me too. I'll try and investigate tonight. 

Do you just have the single zone/alarm group?

Does the mobile app work?

Do you still get the same errors if you perform a set mode?

**Comment by mcgurdan** [comment]
In fact, I just tried bad credentials and I get “Failed to authenticate” which is good. So I am getting the auth but when attempting to use any of the components I get the 403, as if my account does not have to retrieve those details from the API?

**Comment by mcgurdan** [comment]
FWIW I use the Yale Home Smart Living app, and I’m running Yale Smart Home Alarm, not the newer Yale Sync system.

**Comment by mcgurdan** [comment]
Do I need to do anything with the _YALE_AUTH_TOKEN in the script? Do I need to manually get one or is this a universal token?

**Comment by laf** [comment]
> FWIW I use the Yale Home Smart Living app, and I’m running Yale Smart Home Alarm, not the newer Yale Sync system.

Mines the Smart Home Alarm as well so don't worry about that :)



> Do I need to do anything with the _YALE_AUTH_TOKEN in the script? Do I need to manually get one or is this a universal token?

That's universal so don't worry about that.

**Comment by mcgurdan** [comment]
I tried this via PowerShell and using a simple request:

$headers = @{"Content-Type" = "application/x-www-form-urlencoded"; "Authorization" = "Bearer myBearerTokenBlahBlahBlah"}
$url = "https://mob.yalehomesystem.co.uk:6013/yapi/api/panel/mode/"
Invoke-RestMethod -Method Get -Uri $url -Headers $headers

I got the following result:
result  : False
code    : 998
message : no data found!
data    :
time    : 0.0038

It seems very odd, I'm wondering if there is something I need to set on up the yale side? Are there any pre-requisites I might be missing?

**Comment by laf** [comment]
`$url = "mob.yalehomesystem.co.uk:6013/yapi/api/panel/mode"` should have a trailing `/` like:

`$url = "mob.yalehomesystem.co.uk:6013/yapi/api/panel/mode/"`

You shouldn't need to do anything on the yale / app side.

**Comment by mcgurdan** [comment]
The url does have a trailing /

**Comment by mcgurdan** [comment]
I’ll setup a proxy tonight and decrypt the call from my phone to see if the behaviour is different to what has been mentioned here

**Comment by p-tech-se** [comment]
I had the same problem with an empty data blob when using the url with port 6013. Removing the port works for me. I also had success using the port 6003.

**Comment by domwillcode** [comment]
Interesting, seems like it could be exactly that, ans/or load balancing.

IIRC the host:port combo is assigned after the initial /o/token request and returned in the response, so we can use this instead of the hard coded URL after login. 

Ideally we'd store this and the refresh token and re-use it until it expires, rather than login every time an endpoint is hit.

**Comment by laf** [comment]
> The url does have a trailing /

It doesn't show because you've not formatted the code:

![image](https://user-images.githubusercontent.com/3941142/49502402-eeb81d00-f86c-11e8-91df-333ffe553429.png)

**Comment by laf** [comment]
> IIRC the host:port combo is assigned after the initial /o/token request and returned in the response, so we can use this instead of the hard coded URL after login.

Where do you see the host:port returned? This is the response I get:

```
{
  "refresh_token": "X",
  "access_token": "Y",
  "scope": "groups write read basic_profile",
  "expires_in": 36000,
  "token_type": "Bearer"
}
```

And the headers returned don't seem to have it either:

```
Cache-Control →no-store
Connection →keep-alive
Content-Length →190
Content-Type →application/json
Date →Wed, 05 Dec 2018 09:43:25 GMT
Pragma →no-cache
Server →nginx/1.14.0
Vary →Authorization, Cookie
X-Frame-Options →SAMEORIGIN
```

**Comment by p-tech-se** [comment]
The yapi URL is available in the response to a call to the services endpoint. I have updated my fork with a call to services to fetch the URL which is then used later on when fetching status etc.

**Comment by laf** [comment]
@p-tech-se Ace. Can you submit a pull request here so we can get it tested and merged in?

**Comment by p-tech-se** [comment]
Yep, I have created a PR. I hope I got it right, never done a PR before.

**Comment by laf** [comment]
Awesome work :)

@mcgurdan You could test #6 if you like. Let me know if you need help in getting the updated file to replace the current version.

**Comment by mcgurdan** [comment]
Awesome thanks guys, I’ll have a look tonight

**Comment by mcgurdan** [comment]
Awesome, I just tested querying the current status (SSH over my mobile isn’t the easiest but worked lol) - looks like its all working for me! I can test a little further but changes look good so far. Great spot p-tech!

**Comment by domwillcode** [comment]
Thanks for all your hard work guys. PR is in HA for v0.1.6 https://github.com/home-assistant/home-assistant/pull/19495.

**Comment by domwillcode** [comment]
Fixed in #6 and #7.
